htdocs/public/stripe/ipn.php

```diff
@@ -0,0 +1,335 @@
+<?php
+
+define("NOLOGIN",1);		// This means this output page does not require to be logged.
+define("NOCSRFCHECK",1);	// We accept to go on this page from external web site.
+
+$entity=(! empty($_GET['entity']) ? (int) $_GET['entity'] : (! empty($_POST['entity']) ? (int) $_POST['entity'] : 1));
+if (is_numeric($entity)) define("DOLENTITY", $entity);
+
+$res=0;
+if (! $res && file_exists("../../main.inc.php")) $res=@include("../../main.inc.php");		// to work if your module directory is into a subdir of root htdocs directory
+if (! $res) die("Include of main fails");
+
+if (empty($conf->stripeconnect->enabled)) accessforbidden('',0,0,1);
+require_once DOL_DOCUMENT_ROOT.'/user/class/user.class.php';
+require_once DOL_DOCUMENT_ROOT.'/includes/stripe/init.php';
+require_once DOL_DOCUMENT_ROOT.'/stripe/class/stripe.class.php';
+//dol_include_once('/stripeconnect/class/paymentfee.class.php');
+require_once DOL_DOCUMENT_ROOT.'/core/class/ccountry.class.php';
+require_once DOL_DOCUMENT_ROOT.'/commande/class/commande.class.php';
+require_once DOL_DOCUMENT_ROOT.'/compta/paiement/class/paiement.class.php';
+require_once DOL_DOCUMENT_ROOT.'/compta/facture/class/facture.class.php';
+require_once DOL_DOCUMENT_ROOT.'/compta/bank/class/account.class.php';
+require_once DOL_DOCUMENT_ROOT.'/societe/class/societe.class.php';
+
+// You can find your endpoint's secret in your webhook settings
+if (isset($_GET['connect'])){ 
+if (isset($_GET['test']))
+{
+$endpoint_secret =  $conf->global->STRIPE_TEST_WEBHOOK_CONNECT_KEY;
+}
+else 
+{
+$endpoint_secret =  $conf->global->STRIPE_LIVE_WEBHOOK_CONNECT_KEY;	
+}
+}else {
+if (isset($_GET['test']))
+{
+$endpoint_secret =  $conf->global->STRIPE_TEST_WEBHOOK_KEY;
+}
+else 
+{
+$endpoint_secret =  $conf->global->STRIPE_LIVE_WEBHOOK_KEY;
+}
+}
+$payload = @file_get_contents("php://input");
+$sig_header = $_SERVER["HTTP_STRIPE_SIGNATURE"];
+$event = null;
+
+$error = 0;
+    
+try {
+$event = \Stripe\Webhook::constructEvent($payload, $sig_header, $endpoint_secret);
+}                                                                            
+catch(\UnexpectedValueException $e) {
+  // Invalid payload
+  http_response_code(400); // PHP 5.4 or greater
+  exit();
+} catch(\Stripe\Error\SignatureVerification $e) {
+  // Invalid signature
+  http_response_code(400); // PHP 5.4 or greater
+  exit();
+}
+
+// Do something with $event
+
+http_response_code(200); // PHP 5.4 or greater
+$langs->load("main");	
+$user = new User($db);
+$user->fetch(5);
+$user->getrights();
+
+if (! empty($conf->stripeconnect->enabled) && ! empty($conf->multicompany->enabled)) {

```

Comment:
@eldy i try to see on it tonight and tomorrow. For a connect standard account it's not a probleme because already user Oauth.

the probleme is for the custom or express account which only use API REST to get the account. 
we need to save the stripe account too for retrieve the stripe account balance or some additionnal operation
-> save them in llx_const ?

Reply:
Fix done but not realy functionnal/optimum if entity!=1 and not use stripe connect. Use and controle of stripe account is important  for a better security and stripe evolution of API

with new european rules in 2018/2019 PSD2, the use of stripe source for authenticate payment need the use of webhook for the feedback and  failing or dispute action

Reply:
@ptibogxiv 
what is the problem with multicompany ?

Reply:
@hregis no pb with your module 😉. Just stripe webhook request can't include this data in url and for action/payment we need this for use source payment (new European rules for authentificate payment or extra stripe payment gateway as sepa direct debit) and do correct automatic action in dolibarr linked with payment/payout/dispute/fee/refund. So an efficient and easy save of the couple stripe_account/entity is necessary whatever dolibarr only use stripe with your multicompany module (with same or different stripe account for each entity)  or enhance dolibarr+multicompany with my stripeconnect module who transforme dolibarr/stripe in marketplace (the first entity become the unique admin of stripe and can get some extra fee on each payment)

Reply:
@ptibogxiv 
you can not use "metadata" for send the entity ID ??

Reply:
@hregis no :'( stripe don't send it for all their callback ex: 3D secure, payout, failure
It won't be so simple :D
In fact we can, for some webhook callback, retrieve this metadata with some very complexe REST requests but it will be more loud (very loud) and slow that just save the trio stripe_account/entity/mode in the internal dolibarr's database. It's add a complementary security with bad request too.

Reply:
@ptibogxiv 
I do not understand, if it's callback is that you make a request? and if you make a request you can send metadata!? Stripe must return these metadata to you during callback no?

Reply:
@ptibogxiv 
each entity must be able to be autonomous and not depend on the first entity

Reply:
@hregis  
-yes but the callback for some don't return this directly. For stripe request we need the stripe_account (acc_xxxxxxxxxxxxx). All connect's callback has an stripe_account data in the header of the body response so it's more easy and direct to use it. 
ex payment with 3D secure
create 3Dsecure charge not charge but create a secondary card -> callback from stripe -> dolibarr create charge with this ->stripe return if it's ok.
-some automatic stripe's callback as payout have no metada in the response
-some API request in mode >1 entity and stripe not a principal account need include the stripe_account data to cal it.
- we must be compatible with standard an autonomous setting of stripe for entity>1. For the moment, webhook call isn't fully compatible for autonomous stripe entity and a standard account because standard stripe account hasn't stripe_account in the response and for some callback no metadata. 
=>The only solution for use webhook for entity >1 : link another standard stripe account to the principal with Oauth (with or without application fee) or use custom/express stripe account.

with this settings, we can let each entity autonomous or depend of the first entity according to business and/or technical settings of dolibarr.

Reply:
@hregis stripeconnect will support direct charge mode (each entity has each own customer and invoice directly) and separate transfert charge (principal is between customer and each company) for a standard/custom/express connect stripe's account. In dolibarr, all data/action keep autonomous and link to the good entity. It's important for me too.

Reply:
@hregis after reflexion and test, i found solution to solve pb for a standard and alone stripe account.
I will modify ipn.php in order to support both ipn?entity=X and without in connect mode
TODO: create url link  of ipn file to paste in stripe admin on admin/stripe.php